### PR TITLE
[HW3-30-Han-Hua]

### PR DESCRIPTION
**Pull request by hhua**

### CODE REVIEW

File: Lesson3/assignment/answer.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+### 第二题：增加 changePaymentAddress 函数，更改员工的薪水支付地址，思考一下能否使用modifier整合某个功能
+
+
+### 第三题（加分题）：自学C3 Linearization, 求以下 contract Z 的继承线
+- contract O
+- contract A is O
+- contract B is O
+- contract C is O
+- contract K1 is A, B
```
Comment (jonnyhsy) [inline Lesson3/assignment/answer.md]: Note：contract K1 is A, B {...} 这个意思是：合约K1
继承自合约B，合约A，其中B是最先被继承的合约("most-derived"), 继承链应该是K1 <--- B <--- A 见文档：
 - https://solidity.readthedocs.io/en/develop/contracts.html#inheritance  值得好好读一遍

正确的结果：
L(O) = [O]
L(A) = [A, O]
L(B) = [B, O]
L(C) = [C, O]
L(K2) = [K2] + merge(L(C), L(A), [C, A])
      = [K2] + merge([C, O], [A, O], [C, A])
      = [K2, C] + merge([O], [A, O], [A])
      = [K2, C, A, O]
L(K1) = [K1] + merge(L(B), L(A), [B, A])
      = [K1] + merge([B, O], [A, O], [B, A])
      = [K1, B] + merge([O], [A, O], [A])
      = [K1, B, A, O]
L(Z) = [Z] + merge(L(K2), L(K1), [K2, K1])
     = [Z] + merge([K2, C, A, O], [K1, B, A, O], [K2, K1])
     = [Z, K2] + merge([C, A, O], [K1, B, A, O], [K1])
     = [Z, K2, C] + merge([A, O], [K1, B, A, O], [K1])
     = [Z, K2, C, K1] + merge([A, O], [B, A, O])
     = [Z, K2, C, K1, B] + merge([A, O], [A, O])
     = [Z, K2, C, K1, B, A, O]

===== test.py =====  (Note: python的新式类继承用的就是C3算法，所以可以用来测试下)
```python
class O(object):                                                                   
  def __init__(self):                                                              
    print "O"                                                                      

class A(O):                                                                        
  def __init__(self):                                                              
    print "A"                                                                      

class B(O):                                                                        
  def __init__(self):                                                              
    print "B"                                                                      

class C(O):                                                                        
  def __init__(self):                                                              
    print "C"                                                                      

class K1(B, A):                                                                    
  def __init__(self):                                                              
    print "K1"                                                                     

class K2(C, A):                                                                    
  def __init__(self):                                                              
    print "K2"                                                                     

class Z(K2, K1):                                                                   
  def __init__(self):                                                              
    print "Z"


>>> print Z.__mro__
(<class '__main__.Z'>, <class '__main__.K2'>, <class '__main__.C'>, <class '__main__.K1'>, <class '__main__.B'>, <class '__main__.A'>, <class '__main__.O'>, <type 'object'>)
```

File: Lesson3/assignment/yours.sol [hunk 2]
```diff
@@ -1 +1,84 @@
-/*作业请提交在这个目录下*/
+pragma solidity ^0.4.14;
+
+import './Ownable.sol';
```
Comment (jonnyhsy) [inline Lesson3/assignment/yours.sol]: Note: 这里是对的；或者可以采用：
```
import "github.com/OpenZeppelin/zeppelin-solidity/contracts/math/SafeMath.sol";
import "github.com/OpenZeppelin/zeppelin-solidity/contracts/ownership/Ownable.sol";
```
remix会去github取得相应的库文件，本地不需要再拷贝一份重复的。

File: Lesson3/assignment/yours.sol [hunk 3]
```diff
@@ -1 +1,84 @@
...
+
+contract Payroll {
```
Comment (jonnyhsy) [inline Lesson3/assignment/yours.sol]: 让Payroll 继承自Ownable, 可以不用再定义owner state variable和onlyOwner() modifier

File: Lesson3/assignment/yours.sol [hunk 4]
```diff
@@ -1 +1,84 @@
...
+    uint totalSalary;
+    
+    modifier employeeExist(address employeeId) {
+        var employee = employees[employeeId];
+        assert(employee.id != 0x0);
+        _;
+    }
+
+    function _partialPaid(Employee employee) private {
+        uint payment = employee.salary * (now - employee.lastPayday) / payDuration;
```
Comment (jonnyhsy) [inline Lesson3/assignment/yours.sol]: 涉及数值运算的地方最好使用OpenZepplin里的SafeMath.sol库，避免溢出。
文档：
 - https://solidity.readthedocs.io/en/develop/contracts.html#using-for
 - https://github.com/OpenZeppelin/zeppelin-solidity/blob/c7f2138da1b5d588aa1b93b6201b0272f03f3661/contracts/math/SafeMath.sol

File: Lesson3/assignment/yours.sol (hunk 2 above)
Comment (hhua) [reply Lesson3/assignment/yours.sol]: Nice，多谢

File: Lesson3/assignment/yours.sol (hunk 3 above)
Comment (hhua) [reply Lesson3/assignment/yours.sol]: 对，这边忘了用

File: Lesson3/assignment/answer.md (hunk 1 above)
Comment (hhua) [reply Lesson3/assignment/answer.md]: I see. 我把notation搞反了

**Pull request merged**